cupy/cuda/memory_hook.py

```diff
@@ -0,0 +1,96 @@
+import collections
+import threading
+
+thread_local = threading.local()
+
+
+def get_memory_hooks():
+    if not hasattr(thread_local, 'memory_hooks'):
+        thread_local.memory_hooks = collections.OrderedDict()
+    return thread_local.memory_hooks
+
+
+class MemoryHook(object):
+    """Base class of hooks for Memory allocations.
+
+    :class:`~cupy.cuda.MemoryHook` is an callback object
+    that is registered to :class:`~cupy.cuda.SingleDeviceMemoryPool`.
```

Comment:
it was intentional, but now I felt we should provide `device_id` also to memory hook.
I changed the interface of memory hook to do so.